Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift

```diff
@@ -16,130 +16,319 @@ import SymbolKit
+    mutating func applySwiftSyntacticSugar(markers: ContiguousArray<Int>) {
+        assert(!markers.isEmpty, "This is a private helper function and it's the callers responsibility to check if it needs to be called or not.")
+        
+        // Iterating over the UTF-8 string once to find all the balancing angle brackets (`<` and `>`)
+        var markedAngleBracketPairs = ContiguousArray<(open: Int, close: Int)>()
+        markedAngleBracketPairs.reserveCapacity(32) // Some temporary space to work with.
+        
+        var angleBracketStack = ContiguousArray<Int>()
+        angleBracketStack.reserveCapacity(32) // Some temporary space to work with.
+        
+        for index in indices {
+            switch self[index] {
+            case openAngle:
+                angleBracketStack.append(index)
+            case closeAngle where self[index - 1] != hyphen: // "->" isn't the closing bracket of a generic
+                guard let open = angleBracketStack.popLast() else {
+                    assertionFailure("Encountered unexpected generic scope brackets in in \(String(cString: self + [0]))")
+                    return
+                }
+                
+                // Check if this balanced `<` and `>` pair is one of the markers.
+                if markers.contains(open - 1) {
+                    // Save this angle bracket pair, sorted by the opening bracket location.
+                    // Below, these will be iterated over removing the last
+                    if let insertionIndex = markedAngleBracketPairs.firstIndex(where: { open < $0.open }) {
+                        markedAngleBracketPairs.insert((open: open, close: index), at: insertionIndex)
+                    } else {
+                        markedAngleBracketPairs.append((open: open, close: index))
+                    }
+                }
+                
+            default:
+                // Ignore all non `<` or `>` characters
+                continue
+            }
         }
         
-        var result = ""
-        result.reserveCapacity(count)
         
-        var scanner = StringScanner(String(self)[...])
+        assert(markedAngleBracketPairs.map(\.open) == markedAngleBracketPairs.map(\.open).sorted(),
+               "Marked angle bracket pairs \(markedAngleBracketPairs) are unexpectedly not sorted by opening bracket location")
         
-        while let prefix = scanner.scan(until: { $0 == "A" || $0 == "O" || $0 == "D" }) {
-            result += prefix
+        // Iterate over all the marked angle bracket pairs (from end to start) and replace the marked text with the syntactic sugar alternative.
+        while !markedAngleBracketPairs.isEmpty {
+            let (open, close) = markedAngleBracketPairs.removeLast()
+            assert(self[open] == openAngle, "Start marker at \(open) is '\(String(cString: [self[open], 0]))' instead of '>' in \(String(cString: self + [0]))")
```

Comment:
```suggestion
            assert(self[open] == openAngle, "Start marker at \(open) is '\(String(cString: [self[open], 0]))' instead of '<' in \(String(cString: self + [0]))")
```